notes/standardization-meeting-07.12.21.md

```diff
@@ -0,0 +1,284 @@
+
+**Alternatives**
+- ~~Remove the `fun`~~ (makes the grammar ambiguous)
+- Change for Haskell-style `let f = \x y z => body`
+- Add ML-style `let f x y z = body in`. What about partial type annotations?
+
+**In-meeting notes**
+`fun` is fine. In favor of second Ocaml-style function definition `let f x y z =
+...`. What about partial annotations (what is the semantics of `f (x :
+Num) y z =` outside of a typed block) ? The problem is already there with or
+without the OCaml-style function definitions. We should also have an Haskell-style annotation "by pieces" consistent with the let-block syntax:
+
+```nickel
+let f : Foo -> Bar,
+    f x y = ...,
+    g : Bar -> Stuff,
+    g z x = ...,
+in
+```
+
+### Type hole syntax
+[#226](https://github.com/tweag/nickel/issues/226)
+
+Currently, one has to write explicitly type signature to turn typechecking on:
+`let foo : SomeType = ... in ...`
+
+**Proposal**
+`let foo : _ = ... in ...`. Each type hole would be substituted for a
+unification variable at and resolvesd at typechecking. Unconstrained variables
+are substituted for `Dyn`.
+
+**Usages**
+ - typing libraries:
+ ```nickel
+ {
+   bar : Num -> Num = fun x => x + 1,
+   foo : forall a. a -> -a = fun x => x,
+ } : _
+ ```
+ (but beware of duplicated contract checks)
+ - enable typechecking without having to know the top-level type. For debugging.
+ ```
+ let somethingWrong = ...
+ // transform to
+ let somethingWrong : _ = ...
+ ```
+
+**Question**
+Should this generate a contract? Yes.
+
+**In-meeting notes** In favor of typed holes.
+
+Haskell uses the term and syntax of "typed holes" for
+something a tad different (basic interactive programming). If we do so at some
+point, we should use a different character like `?`. `_` for libraries may
+duplicate contracts.  Nonetheless, we should treat this contract duplication
+issue separately: the general issue of optimizing and eliding contracts.
+
+### File extension
+[#357](https://github.com/tweag/nickel/issues/357)
+
+Currently `.ncl` seems to clash with other file formats.
+
+**Proposals**
+Should we have multiple extensions? Probably not (searchability).
+ - `.nickel`
+ - `.nkl`
+ - `.nikl`
+ - `.nicl`
+
+**In-meeting notes**
+No stake. Doesn't impose constraints on other aspects of the development. No real
+objective criterion. We can decide this later.
+
+### Merging with siblings
+Once overriding lands, we may want to refer to fields that will only be defined
+later.
+
+Example:
+```nickel
+{foo = bar + 1} & {bar}
+
+// same name
+{foo = foo + 1} & {foo = 2}
+
+// Encoding CUE if
+let value = {
+  protocol | <http, telnet>
+           | default = `http,
+} & {proxProtocol = protocol}
+```
+
+But it introduces dynamic scoping, which is bad.
+
+**Proposal**
+- add it nevertheless
+- force the definition of an interface:
+  ```nickel
+  {foo = bar + 1, bar | Num} & {bar}
+  // For more complex/dynamic interfaces
+  let Interface = {bar | Num} in
+  {foo = bar + 1, ..Interface}
```

Comment:
Does this have to be different to something like the following because scope would be calculated before the contract would be verified, and would therefore rely on dynamic scoping?
```
  let Interface = {bar | Num} in
  {foo = bar + 1} | #Interface
```

I'm worried this may feel a little redundant in a case like this:
```
  let Interface = {bar | Num} in
  let impl | #Interface = {foo = bar + 1, ..Interface},
  ..
```

Reply:
>Does this have to be different to something like the following because scope would be calculated before the contract would be verified, and would therefore rely on dynamic scoping?

At first sight, I would say so. Although we could also have rule like if `Interface` is statically known to be a record with field `bar`, then `impl | #Interface` would be sufficient. We would probably want a similar constraint to allow the user to write `..Interface` anyway, so that wouldn't be very different. In the end, I think the idea to avoid the pitfalls of dynamic scoping is that `Interface` has enough static information that we can know where `bar` comes from and that it isn't any random unbound identifier.

Reply:
The problem with this kind of rules though, is that they are often fragile, as they are a bit ad hoc. Any indirection or dynamic contract would break the code that was working in simple cases.

Reply:
Yep, makes sense, thank you.